### PR TITLE
Add UI 2.0 version of Character UI

### PR DESCRIPTION
**Pull request by Mehuge**
Migrated UI 2.0 version of character UI from cu-ui-v2.  

Please be aware of this issue: https://forums.camelotunchained.com/topic/12185-invariant-violation-namerender-a-valid-reactcomponent-must-be-returned-you-may-have-returned-undefined-an-array-or-some-other-invalid-object/

It may affect others who try and build this project.

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/CUModSquad/cu-ui/pull/2" target="_blank">ReviewNinja

**Comment by Sschlenker** [comment]
More of a conversation point than anything.  note: I added a star as I did not see anything that stood out to me.

Back to the conversation point.  Are we not necessarily required to use typescript markup in the tsx setups?   I know it has been generically stated that as long as the end result is valid working es6 then no biggie.  Think I also heard that JB was not going to be super forceful on it, but was not sure about the official ui stuff.

**Comment by Mehuge** [comment]
I wrote these before any of that had been decided.  I wanted to convert them as they are before any attempt to convert them to typescript classes.  

I did trial run doing that and there are a few issues that cropped up, one of those was mixins, used by the component to hook store actions up to the component.  That needs to be done differently, and don't know the best way to do that yet (open to suggestions).  I have seen a few different suggestions, none of which stood out as yeah, that's the way to go, though haven't looked into it in any depth.

I would rather treat the migration to cu-ui separate to any conversion to es6.  We will need to do something eventually as there are a few deprecated calls being made.

**Comment by Mehuge** [comment]
I wrote these before any of that had been decided. I wanted to convert them as they are before any attempt to convert them to typescript classes.

I did trial run doing that and there are a few issues that cropped up, one of those was mixins, used by the component to hook store actions up to the component. That needs to be done differently, and don't know the best way to do that yet (open to suggestions). I have seen a few different suggestions, none of which stood out as yeah, that's the way to go, though haven't looked into it in any depth.

I would rather treat the migration to cu-ui separate to any conversion to es6. We will need to do something eventually as there are a few deprecated calls being made.

File: character/package.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "type": "git",
+    "url": "https://github.com/csegames/cu-ui.git"
+  },
+  "dependencies": {
+    "cu-components": "^0.1.2",
+    "cu-core": "^0.1.3",
+    "cu-events": "^0.1.5",
+    "cu-stores": "^0.1.5"
+  },
+	"peerDependencies": {
```
Comment (Mehuge) [inline character/package.json]: These snuck in (peerDependencies) whilst experimenting, don't think they are needed.

Comment (Mehuge) [inline character/package.json]: They should probably go back into dependencies.

File: character/src/ts/main.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        health={character.health} maxHealth={character.maxHealth}
+        stamina={character.stamina} maxStamina={character.maxStamina} />
+		);
+	}
+});
+
+events.on("init", function() {
+  character.actions.start();          // HACK: for cuAPI time limited events issue
+  React.render(<Character/>, document.getElementById("cse-ui-character"));
+});
```
Comment (codecorsair) [inline character/src/ts/main.tsx]: missing a new line at the end of the file.

File: character/src/sass/main.scss [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    /*
+     * These styles overide the cbase unitframe.css styles defined by
+     * ../cu-lib/views/unitframe.css.  For the most part this means
+     * different position and sizes of the bars, and different artwork
+     * for the frame and bars.  And some different colours.  The fonts
+     * and other stuff are inherited from unitframe.css
+    */
+
+    #unitframe {
```
Comment (codecorsair) [inline character/src/sass/main.scss]: indentation should be should be 2 spaces not 4.

File: character/src/character.ui [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "character",
+  "mainFile": "index.html",
+  "rect": {
+      "left": 0,
+      "top": 0,
+      "right": 245,
+      "bottom": 125
```
Comment (codecorsair) [inline character/src/character.ui]: 2 space indentation

File: character/src/ts/main.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+const Character = React.createClass({
+
+  // Hook store up to component.  Each time character data is changed,
+  // our state is updated, triggering a render
+  mixins: [
+    Reflux.connect(character.store, 'character')
+  ],
+
+	// Provide an initial state (TODO: is there a better way to do this?)
+	getInitialState: function() {
```
Comment (codecorsair) [inline character/src/ts/main.tsx]: indentation.

**Comment by codecorsair** [comment]
Good aside from the formatting errors.  Once those are fixed up you can merge it. :+1:

**Comment by codecorsair** [comment]
> Are we not necessarily required to use typescript markup in the tsx setups? I know it has been generically stated that as long as the end result is valid working es6 then no biggie. Think I also heard that JB was not going to be super forceful on it, but was not sure about the official ui stuff.

In regard to this question, we will be strict on enforcing the use of TypeScript.  ES6 style is to be used in all cases except where a mixin is needed for React Components at this time.  I'll be looking into a workaround for using reflux without a mixin, at that time we will then be fully using ES6 style TypeScript.

**Comment by codecorsair** [comment]
Good to go! :+1:

**Pull request merged**